compiler/src/iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -2275,10 +2275,89 @@ struct ElideNoOpAsyncExecuteOp : public OpRewritePattern<AsyncExecuteOp> {
   }
 };
 
+struct DeduplicateYieldCmdExecuteOp : public OpRewritePattern<AsyncExecuteOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AsyncExecuteOp op,
+                                PatternRewriter &rewriter) const override {
+    llvm::SmallVector<bool> keepYield;
+    llvm::SmallVector<Value> yieldOperands;
+    llvm::SmallVector<int> remapping;
+
+    auto yield =
+        cast<IREE::Stream::YieldOp>(op.getBody().front().getTerminator());
+    int64_t oldYieldCount = yield.getResourceOperands().size();
+    for (int i = 0, s = oldYieldCount; i < s; ++i) {
+      auto operand = yield.getResourceOperands()[i];
+
+      auto find =
+          std::find(yieldOperands.begin(), yieldOperands.end(), operand);
+      if (find != yieldOperands.end()) {
+        keepYield.push_back(false);
+        remapping.push_back(find - yieldOperands.begin());
+        continue;
+      }
+
+      remapping.push_back(yieldOperands.size());
+      keepYield.push_back(true);
+      yieldOperands.push_back(operand);
+    }
+
+    if (oldYieldCount == yieldOperands.size()) {
+      return failure();
+    }
+
+    llvm::SmallVector<Type> newTypes;
+    llvm::SmallVector<Value> newResultSizes;
+
+    for (int i = 0; i < oldYieldCount; ++i) {
+      if (!keepYield[i])
+        continue;
+      newTypes.push_back(op.getResults()[i].getType());
+      newResultSizes.push_back(op.getResultSizes()[i]);
+    }
+
+    auto newExecuteOp = rewriter.create<IREE::Stream::AsyncExecuteOp>(
+        op.getLoc(), newTypes, newResultSizes, op.getAwaitTimepoint(),
+        op.getResourceOperands(), op.getResourceOperandSizes(),
+        llvm::SmallVector<int64_t>());
```

Comment:
we'll need to preserve tied operands - this is stripping them all

(always useful to throw in a tied operand to ensure they are not stripped, like by adding a `..., %tied) -> %tied` to the end so that it's clear the indices are updated properly - updating index 0 tied operands is usually too easy :)